### ISSUE TITLE
Keep a pool of easy-to-solve issues for beginners

### ISSUE THREAD

**Issue by marcusfuchs** [RWTH-EBC/AixLib#372]
## What is the problem?

- New contributors (e.g. students using AixLib for their thesis work) sometimes struggle with learning Modelica, AixLib, Git and the necessary workflows all at the same time while already being occupied with working on their thesis topic. 

## Why do we want to solve it?

- If we help new contributors to quickly learn how to contribute to AixLib, things may get easier for all of us.

## How do we want to solve it?

- I propose to always keep a pool of very easy-to-solve issues (maybe 3 or 4 of them) for beginners to start contributing to AixLib. This could be small issues like correcting a typo, reformating or adding a bit of documentation etc. 
- These issues could be labeled to make them easy to find. [Facebook's React-devtools repo](https://github.com/facebook/react-devtools/issues) is a nice example for this approach. They use the labels `difficulty:starter` and `good first bug`. 
- If new contributors start with solving one if these easy issues before they begin their own development, I would expect that we will run into a lot less trouble later on (e.g. when today students work on their models for half a year locally and are then told to "just put your models into AixLib")

So, what do you guys think about such an approach? If there are no objections, I will try to add one or two of these easy-to-solve issues.

------------
### Update

I added the `good first issue` label and created a first example issue at #373.

**Comment by PRemmen** [comment]
No objections at all. This might help to get new user actively involved into our development process!

**Comment by marcusfuchs** [comment]
I think 2 thumps-up are motivation enough, so I'll add a first example issue that is very easy to solve.

**Comment by marcusfuchs** [comment]
I have added [a short tutorial on how to solve a first issue for beginners to our Wiki](https://github.com/RWTH-EBC/AixLib/wiki/First-contribution).

**Comment by marcusfuchs** [comment]
@alexanderAKU and @aku-cbe: 
As you are planning to work intensively on AixLib as described in #360 and you would both be first-time contributors to AixLib: would you be interested in having a look at the [short tutorial on how to solve a first issue for beginners to our Wiki](https://github.com/RWTH-EBC/AixLib/wiki/First-contribution) and try this process?

**Comment by alexanderAKU** [comment]
We are going to try the [short tutorial on how to solve a first issue for beginners in our Wiki](https://github.com/RWTH-EBC/AixLib/wiki/First-contribution).

**Comment by aku-cbe** [comment]
Is there a difference in pushing branches and pushing commits?
In the tutorial first contribution it says I should push my commits regularly, and in the last section I should now finally push my branch to the origin. 
But I pushed my branch already together with the commits, haven't I ?
So this statement doesn't make any sense to me.

@marcusfuchs

**Comment by marcusfuchs** [comment]
@aku-cbe Thanks for the feedback. I slightly changed the wording, so that you are now asked not to push commits, but your branch containing the commits. Also, in the end you are asked to push your branch *again* after merging the current `development` branch from origin. Is that clearer?

**Comment by aku-cbe** [comment]
Yes I guess.
But I started a Pull Request to merge into master, not development. 
Can I delete that somehow? :(

**Comment by marcusfuchs** [comment]
@aku-cbe I already fixed that. I changed the tutorial to mention changing the base branch specifically, so it will be easier for the next persion. In addition, I mentioned adding your author information to the model's revision history. Thanks again! Only by actually trying the process we learn what is missing from the tutorial.

**Comment by aku-cbe** [comment]
@marcusfuchs asked me to summarize my experience with this process.#

It first I was quite reluctant to do this extra work because, after the tutorial I did, I wanted to start with my work.
In the end it took me maybe 1 to 2 hours of work and I learned quite a bit about the issue structure and the different branches, i.e. the architecture of the AixLib.
I guess this will proof helpful when starting my own development in the AixLib.
The tutorial was mostly clear and well stated, especially after the latest changes and updates. 
On the other hand, the tasks are so small that you might be faster changing it personally than to write an issue about such things, and the tutorial for itself is quite good in itself.

So yes, me personally it helped, but you have to decide which amout of effort you want to put in creating these easy first issues.

**Comment by marcusfuchs** [comment]
@aku-cbe Thank you for the open feedback! And I'm happy to hear that it helped you. For me, the efforts were not that significant:

- Finding a small error in the library and creating the issue: 10-15 minutes
- 1st review: around 10 minutes
- 2nd check and final merge: 5 minutes

All of that is low-commitment-work that can be done in between other tasks. And I would expect that this time investment will save us a lot of time on larger reviews when first-time-contributors are already familiar with our process rather than learning about it when trying to merge large pieces of work.

So if anyone has a better idea on how to get first-time-contributors familiar with our workflows I'd be happy to try it. If not, I suggest to try and continue with the `first good issue` approach.

**Comment by MirkoEngelpracht** [comment]
@marcusfuchs Marcus asked me to share my experience with the tutorial and the merging process. Therefore, I'll give a short feedback:

**The tutorial**
In my opinion, the tutorial is easy to follow and most of the instructions are clearly understandable. However, I've struggled a bit with two instructions:

> Make sure your branch starts off the most recent version of the development branch.

At the beginning, I was not sure how to do this because I thought all new branches will start of the master branch. In the end, I first cloned the whole AixLib and then switched to the development branch. Afterwards, I created my issue branch and noticed that it started of from the development branch. Is that the right thing to do? Or do I have to do something else?

> Before you do so, make sure to pull the latest version of the development branch to your local branch.

After finishing my issue working on my issue branch, I switched back to the development branch and used the pull command to get the latest version. Afterwards I switched back to my issue branch and used the push command to upload it. Is that the right thing to do? I ask that because I thought that I would just update the development branch and not my issue branch by using this order. Or did I miss something?

**The first issue**
The first issue was easy to solve and it has helped me to get started with Modelica/Dymola as well as to understand the work flow in general. Therefore, I think it is a good thing :)


All in all, this helped me a lot to understand your work flow and I guess that it will be now much easier to start with my thesis. Thank you very much for this.

**Comment by marcusfuchs** [comment]
@MirkoEngelpracht Thank you for the feedback! I updated both parts of the tutorial which were not clear to you. Could you please check if it is clearer now? (hint: You correctly branched off the development, but you did not merge the most recent development back into your local branch before the pull request. This is no problem as long as the development has not changed since you branched off it, but it could lead to problems when you work on your branch for longer)

**Comment by MirkoEngelpracht** [comment]
@marcusfuchs Thank you for the update. It's clearer now!

**Comment by marcusfuchs** [comment]
Ok perfect, thanks again. I will close this issue for the moment, but feel free to reopen any time to continue the discussion or open a new one for specific feedback to parts of the tutorial or the process. 

As a conclusion, I will try to continue supporting the good first issue process and encourage all students and new colleagues to go through the tutorial before starting bigger tasks.
